core/state.go

```diff
@@ -362,39 +352,97 @@ func (s *State) Class(classHash *felt.Felt) (*DeclaredClass, error) {
 	return &class, nil
 }
 
-// updateContractStorage applies the diff set to the Trie of the
-// contract at the given address in the given Txn context.
-func (s *State) updateContractStorage(stateTrie *trie.Trie, addr *felt.Felt, diff []StorageDiff, onChanged OnValueChanged,
-	blockNumber uint64,
-) error {
-	contract, err := NewContract(addr, s.txn)
+func (s *State) updateStorageBuffered(contractAddr *felt.Felt, updateDiff []StorageDiff, blockNumber uint64, logChanges bool) (
+	*db.BufferedTransaction, error,
+) {
+	// to avoid multiple transactions writing to s.txn, create a buffered transaction and use that in the worker goroutine
+	bufferedTxn := db.NewBufferedTransaction(s.txn)
+	bufferedState := NewState(bufferedTxn)
+	bufferedContract, err := NewContract(contractAddr, bufferedTxn)
 	if err != nil {
-		if !errors.Is(err, ErrContractNotDeployed) {
-			return err
+		return nil, err
+	}
+
+	onValueChanged := func(location, oldValue *felt.Felt) error {
+		if logChanges {
+			return bufferedState.LogContractStorage(contractAddr, location, oldValue, blockNumber)
 		}
+		return nil
```

Comment:
> unnecessary branching

I think it looks better if it is in the callback. Also, that branch is almost entirely static, so it will always be correctly predicted by the processor. So it has near-0 performance impact.